spa_client_concerns.md

```diff
@@ -0,0 +1,29 @@
+
+Okay, so all of this is why you were assigned a task to build a new web app with the latest front-end techniques that relies on OAuth 2 APIs. Now you've done it, so let's start attacking it.
+
+First and foremost, we have to identity what can be exploited in this environment. Static JavaScript code that doesn't make any APIs calls is essentially attack-proof - the worst thing an attacker could do is mess up his own browser.
+
+It is the API usage that we have to look very closely at. You as the app developer know that your app needs an access token in order to call the APIs. You are faced with a number of implementation choices in order to get the token. If you make the wrong choices, then I can successfully attack your system. My end goal is to be able to call those APIs in ways that you don't want me to.
+
+In OAuth 2, the entity which obtains access tokens and uses them to make bearer requests is called the "client". In this case, it's the single page app code running in the user's browser that is the client. The first choice the client developer has to make is how to get the tokens.
+
+The first and simplest mistake you might have made was to register your SPA client with the authorization server (AS) in the same way as a client that runs in a server context. Server-side clients are called "confidential clients" - they have an id and a secret, so they can be identified easily and reliably in the calls made to to the AS. This works for server-side clients because they never expose the client secret.
+
+You might think your client can operate similarly - just put the client id and secret in the JavaScript code for your app. However, that code is sent to my browser when I use your application. I can easily review the code (or the network requests made by the code) and find that client secret. Suddenly it's not so secret, and now that I have it I can fool the AS into thinking that my attack code is really your legitimate client.
+
+To prevent this, you need to register your app as a "public client". This means it has no client secret, and instead relies on other means to reliably identify the legitimate client. Specifically, it relies on a specific, registered "redirect_uri" that points to your SPA. For a mere redirect uri to be sufficient, it should also only use the HTTPS protocol; otherwise, I could set up a man-in-the-middle system and attack your users by intercepting their traffic (stealing their access tokens in the process).
```

Comment:
"stealing their access tokens in the process"—maybe not narrow it down to tokens, rather "transmitted credentials" or something. We, in fact, do not recommend Implicit Grant in the end.